File: Difficulty/Hard/[52] N-Queens II/solution.cpp
class Solution {
public:
    int totalNQueens(int n) {
        int res = 0;
        if (n < 1) return res;
        vector<string> puz(n, string(n, '.'));
        vector<vector<int>> positions;
        nQueens(res, puz, 0, n, positions);
        return res;
    }

    void nQueens(int& res, vector<string>& puz, int row, int n, vector<vector<int>>& positions){
        if (puz[n-1] != string(n, '.')){
            res++;
            return;
        }
        for (int col = 0; col < n; col++){
            int posId = 0;
            for (auto& pos : positions){
                if (abs(pos[0] - row) == abs(pos[1] - col) || pos[1] == col)
                    break;
                posId++;
            }
            if (posId == positions.size()){
                positions.push_back({row, col});
                puz[row][col] = 'Q';
                nQueens(res, puz, row+1, n, positions);
                puz[row][col] = '.';
                positions.pop_back();
            }
        }    
    }
};

File: Difficulty/Hard/[52] N-Queens II/solution.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        if n < 1:
            return 0
        self.res = 0
        row, positions, cur = 0, [], [n * ['.'] for _ in range(n)]
        self.Nqueens(cur, row, n, positions)
        return self.res
    
    def Nqueens(self, cur, row, n, positions):
        if cur[n - 1] != n * ['.']:
            list_to_str = []
            for l in cur:
                list_to_str.append(''.join(l))
            self.res += 1
            return
        for col in range(n):
            posId = 0
            for pos in positions:
                if abs(pos[0] - row) == abs(pos[1] - col) or pos[1] == col:
                    break
                posId += 1
            if posId == len(positions):
                positions.append([row, col])
                cur[row][col] = 'Q'
                self.Nqueens(cur, row + 1, n, positions)
                cur[row][col] = '.'
                positions.pop()